src/interfaces.py

```diff
@@ -5,10 +5,32 @@
 
 from pandas import DataFrame
 
-TypedDataFrame = tuple[DataFrame, dict[str, Any]]
+
+class LenCapable(Protocol):
+    """A protocol that enforces the presence of __len__."""
+
+    def __len__(self) -> int:
+        """Return the number of items present in this interface."""
+
 
 # This will represent your data type (DataFrame, dict, etc.)
-T = TypeVar("T")
+T = TypeVar("T", bound=LenCapable)
+
+
+class TypedDataFrame:
+    """A wrapper around (DataFrame, metadata) with a __len__ method."""
+
+    def __init__(self, dataframe: DataFrame, types: dict[str, Any]):
+        self.dataframe = dataframe
+        self.types = types
+
+    def __len__(self) -> int:
```

Comment:
I had to introduce this class so that we would have an accurate __len__ implementation on the fetched results (regardless of their structure).